docs/development.md

```diff
@@ -0,0 +1,775 @@
+capable of interfacing with a given provider. Next, the `getStepStartStates`
+will run to get a list of integration phases and steps that should be executed.
+Finally, the phases in `integrationStepPhases` are then executed in sequence.
+
+### What's in the `IntegrationExecutionContext` and `IntegrationExecutionStepContext`?
+
+The `IntegrationExecutionContext` contains some basic utilities to provide some
+information about the entity data.
+
+It includes a `logger` that can be used to display additional information
+throughout an integration run. An `instance` field containing the integration
+instance, which contains configuration field values stored under a `config`
+field.
+
+The `IntegrationExecutionStepContext` contains the same data stored in the
+`IntegrationExecutionContext` but also contains a `jobState` object that
+provides utilities for collecting and validating graph data. More details about
+the `jobState` is detailed in the [Data collection](# Data collection) section
+below.
+
+### Additional utilities
+
+#### Graph model validation
+
+To assist with keeping data consistent with our graph data model, the
+integration SDK will expose utility functions (`createIntegrationEntity` and
+`createIntegrationRelationship`) for validating entities and relationships based
+on their assigned `_class`.
+
+These functions will automatically validate that an entity contains required
+fields.
+
+#### Raw data collection
+
+In addition to performing validation, the `createIntegrationEntity` and
+`createIntegrationRelationship` will also automatically encode and store the
+original provider data under a `__rawData` field. For now, this will always
+assume that data coming in was stored as `json`, but support for other data
+types will come later.
+
+### Auto-magic behind the framework
+
+#### Logging
+
+When running an integration via the CLI tool will automatically log out
+information about the state changes that the integration. This allows for
+developers to keep track of what phase and step an integration is performing
+without the need to explicily add logging information themselves.
+
+#### Data collection
+
+The `executionContext` that is provided in the `executionHandler` step exposes a
+`jobState` utility that can be used to collect entity and relationship data via
+`addEntities` and `addRelationship` functions. The `jobState` utility will
+automatically flush the data to disk as a certain threshold of entities and
+relationships is met. The data flushed to disk are grouped in folders that based
+on the phase and step that was run. Flushed data will be replicated to a
+separate directory that groups data by an entity or relationship's `_type`.
+Additional utilities for reading in entities or relationships via the `_type`
+are provided on the `jobState` object (the `listEntitiesByType` and
+`listRelationshipsByType` functions). From our experience, integrations most
```

Comment:
I love that we start with just `_type` searching. I think that we should consider making these functions `iterateEntities({ _type: "some_type" }, (e) => { ... })`, with that input being `type IterateEntitiesFilter = { _type: string }` at first, and later ` type IterateEntitiesFilter = { _type?: string, _class?: string | string[], [property: string]: value: undefined | string | RegExp | number | whateverapropcanbeingraph }`. This will make the interface consistent no matter what properties we search by in the future. Same for `iterateRelationships()` of course.

Reply:
Seems reasonable. Will change